File: zachita_information/lab6.py
import math
import random

# вычисление обратного числа по модулю n 
# нужно для вычисления d
def inv(a, n): 
  if (a % n) != 0:
    result = 1
    while True:
      if ((result * a) % n) == 1:
        return result
      else:
        result += 1
  else:
    return None

# зашифрование
def encrypt(message, keys):
  pk = keys[0]
  e = pk[0]
  n = pk[1]
  result = (message ** e) % n
  return int(result)

# дешифрование
def decrypt(message, keys):
  
  sk = keys[1]
  n = sk[1]
  d = sk[0]
  result = (message ** d) % n

  return int(result)


def main():

    print("\nRSA\n")

    p = 13
    q = 17

    # тут вычисляем ключи
    n = p * q

    ############################# ПЕРВЫЙ КЛЮЧ
    # эйлер
    eiler = (q - 1) * (p - 1)

    # от нуля до произведения p*q
    e = random.randint(3, n - 1)

    gcd = math.gcd(e, eiler)

    while gcd != 1:
        e = random.randint(3, n - 1)
        gcd = math.gcd(e, eiler)

    ############################# ВТОРОЙ КЛЮЧ
    d = inv(e, eiler)
    if d is None:
        return print("error")

    # сами ключи
    pk = (e, n)
    sk = (d, n)

    keys = pk, sk

    print("Ключи: ", keys)

    message = 14
    print("Исходное сообщение: ", message)

    enc = encrypt(message, keys)
    print("Зашифрованное сообщение: ", enc)

    dec = decrypt(enc, keys)
    print("Разшифрованное сообщение: ", dec)

    print()


if __name__ == '__main__':
  main()

File: zachita_information/lab6.cpp
#include <iostream>
#include <cmath>
#include <random>

using namespace std;

//вычисление обратного числа по модулю для вычисления d
int inv(int a, int n) 
{
    if (a % n != 0) 
    {
        int result = 1;

        while (true) 
        {
            if ((result * a) % n == 1) 
            {
                return result;
            }
            result += 1;
        }
    }
    else {
        return -1;
    }
}

//функция нахождения нод двух чисел
int gcd_func(int a, int b)
{
    if (a < b) 
    {
        swap(a, b);
    }
    while (a % b != 0) 
    {
        a = a % b;
        swap(a, b);
    }
    return b;
}

//зашифровка сообщения
int zachifr(int message, int pk[])
{
    int e = pk[0];
    int n = pk[1];

    return fmod((pow(message, e)), n);
}

//дешифрование
int dechifr(int message, int sk[]) 
{
    int d = sk[0];
    int n = sk[1];

    return fmod((pow(message, d)), n);
}



int main() 
{

    setlocale(LC_ALL, "Russian");

    cout << "RSA" << endl;

    int p = 13;
    int q = 17;

    int n = p * q;

    int eiler = (q - 1) * (p - 1);

    //генерируем е рандомно от нуля до n 
    int e = rand() % (n);

    //находим нод
    int gcd = gcd_func(e, eiler);

    
    while (gcd != 1) 
    {
        //e = rand() % (n);
        gcd = gcd_func(e, eiler);
    }

    //////второй ключ

    int d = inv(e, eiler);
    if (d == -1) 
    {
        cout << "error" << endl;
        return 0;
    }

    //полученные ключи
    int pk[] = { e, n };
    int sk[] = { d, n };

    cout << "Ключи: " << endl;

    cout << "PK = [" << pk[0] << ", " << pk[1] << "];" << endl;
    cout << "SK = [" << sk[0] << ", " << sk[1] << "];" << endl;

    int message = 113;
    cout << "Исходное сообщение: " << message << endl;

    int enc = zachifr(message, pk);
    cout << "Зашифрованное сообщение: " << enc << endl;

    int dec = dechifr(enc, sk);
    cout << "Раcшифрованное сообщение: " << dec << endl;

    return 0;
}
